### PR TITLE
Add modified transducer loss for AIShell dataset

### PR DESCRIPTION
**Pull request by csukuangfj**
Still in training. Will post results later.

[EDITED]:
Training command is
```bash
export CUDA_VISIBLE_DEVICES="0,1,2"

./transducer_stateless_modified/train.py \
  --world-size 3 \
  --num-epochs 60 \
  --start-epoch 0 \
  --exp-dir transducer_stateless_modified/exp \
  --max-duration 300 \
  --lr-factor 2.5 \
  --context-size 2 \
  --modified-transducer-prob 0.25
```

### CODE REVIEW

**Comment by csukuangfj** [comment]
WERs for modified transducer (using --modified-transducer-prob 0.25) with extra data (using --datatang-prob 0.2):

| decoding method | CER on test | comment |
|--|------------|-----------|
| greedy search | 4.94 | --epoch 89 --avg 38 --max-sym-per-frame=1 |
| modified beam search | 4.68 | --epoch 89 --avg 38 --beam-size=4|

Training command is
```bash
cd egs/aishell/ASR
./prepare.sh --stop-stage 6
./prepare_aidatatang_200zh.sh

export CUDA_VISIBLE_DEVICES="0,1,2"

./transducer_stateless_modified-2/train.py \
  --world-size 3 \
  --num-epochs 90 \
  --start-epoch 0 \
  --exp-dir transducer_stateless_modified-2/exp-2 \
  --max-duration 250 \
  --lr-factor 2.0 \
  --context-size 2 \
  --modified-transducer-prob 0.25 \
  --datatang-prob 0.2
```

Decoding command is

```bash
for epoch in 89; do
  for avg in 38; do
  ./transducer_stateless_modified-2/decode.py \
    --epoch $epoch \
    --avg $avg \
    --exp-dir transducer_stateless_modified-2/exp-2 \
    --max-duration 100 \
    --context-size 2 \
    --decoding-method greedy_search \
    --max-sym-per-frame 1
  done
done

for epoch in 89; do
  for avg in 38; do
    ./transducer_stateless_modified-2/decode.py \
    --epoch $epoch \
    --avg $avg \
    --exp-dir transducer_stateless_modified-2/exp-2 \
    --max-duration 100 \
    --context-size 2 \
    --decoding-method modified_beam_search \
    --beam-size 4
  done
done
```

Will update `RESULTS.md` later.

**Comment by danpovey** [comment]
Very nice!  It seems that multi-dataset thing is a useful feature.

**Comment by csukuangfj** [comment]
The following summarizes the results of this PR

## basline (not using extra data)

Tensorboard log: https://tensorboard.dev/experiment/C27M8YxRQCa1t2XglTqlWg/

Pre-trained model, decoding logs and results: https://huggingface.co/csukuangfj/icefall-aishell-transducer-stateless-modified-2022-03-01

Training command

```bash
./transducer_stateless_modified/train.py \
  --world-size 3 \
  --num-epochs 90 \
  --start-epoch 0 \
  --exp-dir transducer_stateless_modified/exp-4 \
  --max-duration 250 \
  --lr-factor 2.0 \
  --context-size 2 \
  --modified-transducer-prob 0.25
```

|                        | test |comment                                                         |
|------------------------|------|----------------------------------------------------------------|
| greedy search          | 5.22 |--epoch 64, --avg 33, --max-duration 100, --max-sym-per-frame 1 |
| modified beam search   | 5.02 |--epoch 64, --avg 33, --max-duration 100  --beam-size 4         |

(Note: It is also trained for 90 epochs, but the best CER is at `--epoch 64 --avg 33`


## with extra data (--datatang-prob 0.2)

Tensorboard log: https://tensorboard.dev/experiment/oG72ZlWaSGua6fXkcGRRjA/

Pre-trained model, decoding logs and results: https://huggingface.co/csukuangfj/icefall-aishell-transducer-stateless-modified-2-2022-03-01

Training command:
```bash
./transducer_stateless_modified-2/train.py \
  --world-size 3 \
  --num-epochs 90 \
  --start-epoch 0 \
  --exp-dir transducer_stateless_modified-2/exp-2 \
  --max-duration 250 \
  --lr-factor 2.0 \
  --context-size 2 \
  --modified-transducer-prob 0.25 \
  --datatang-prob 0.2
```

|                        | test |comment                                                         |
|------------------------|------|----------------------------------------------------------------|
| greedy search          | 4.94 |--epoch 89, --avg 38, --max-duration 100, --max-sym-per-frame 1 |
| modified beam search   | 4.68 |--epoch 89, --avg 38, --max-duration 100  --beam-size 4         |

## with extra data (--datatang-prob 0.1)

Tensorboard log: https://tensorboard.dev/experiment/9yh9j674Rb6BmjpwZlwcEg/

Training command:
```bash
./transducer_stateless_modified-2/train.py \
  --world-size 3 \
  --num-epochs 90 \
  --start-epoch 0 \
  --exp-dir transducer_stateless_modified-2/exp-33 \
  --max-duration 250 \
  --lr-factor 2.0 \
  --context-size 2 \
  --modified-transducer-prob 0.25 \
  --datatang-prob 0.1
```

|                        | test |comment                                                         |
|------------------------|------|----------------------------------------------------------------|
| greedy search          | 4.98 |--epoch 89, --avg 38, --max-duration 100, --max-sym-per-frame 1 |
| modified beam search   | 4.73 |--epoch 89, --avg 38, --max-duration 100  --beam-size 4         |

---

You can see that:
- Using extra data in the training helps to decrease the WER (4.68 vs 5.02)
- Increasing the probability of using extra data (from 0.1 to 0.2) decreases the WER (4.73 vs 4.68)

**Comment by danpovey** [comment]
RE "Increasing the probability of using extra data (from 0.1 to 0.2) decreases the WER (4.73 vs 4.68)"... I think this is too small to be significant.  What I'd say is, it probably indicates that either the optimal prob is somewhere in the range [0.1..0.2], or simply that the outcome is not very sensitive to that probability.
